"""Types for generating pipeline results."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, TypeAdapter

from prp.models.manifest import URI
from prp.parse.models.base import VariantBase
from prp.models.base import RWModel
from prp.parse.models.enums import AnalysisType

SCHEMA_VERSION: int = 2


class MethodIndex(RWModel):
    """Container for key-value lookup of analytical results."""

    type: str
    software: str
    result: Any


class ReferenceGenome(RWModel):
    """Reference genome."""

    name: str
    accession: str
    fasta: str
    fasta_index: str
    genes: str


class IgvAnnotationTrack(RWModel):
    """IGV annotation track data."""

    name: str  # track name to display
    file: str  # path to the annotation file


class QcMethodIndex(RWModel):
    """QC results container.

    Based on Mongo db Attribute pattern.
    Reference: https://www.mongodb.com/developer/products/mongodb/attribute-pattern/
    """

    software: str
    version: str | None = None
    result: Any


class CdmQcMethodIndex(QcMethodIndex):
    """Qc results container for CDM"""

    id: str


class PipelineArtifact(BaseModel):
    """Describe a file generated by the pipeline."""

    software_name: str
    software_version: str
    uri: URI


class PipelineDefinition(BaseModel):
    """Static metadata about the pipeline code itself."""

    name: str
    version: str
    commit: str | None = None
    release_life_cycle: Literal["development", "staging", "production", "unknown"]


class PipelineRunConfig(BaseModel):
    """Configuration used for this specific execution."""

    command: str
    analysis_profile: list[str] = Field(default_factory=list)
    configuration_files: list[str] = Field(default_factory=list)


class PipelineInfo(BaseModel):
    """Full description of the pipeline and its execution."""

    definition: PipelineDefinition
    run_config: PipelineRunConfig
    artifacts: list[PipelineArtifact]


class PipelineRun(BaseModel):
    """Describe a execution of a pipeline."""

    pipeline_run_id: str
    assay: str
    executed_at: datetime
    pipeline_info: PipelineInfo


class SequencingInfo(BaseModel):
    """Information on the sample was sequenced."""

    sequencing_run_id: str
    platform: str
    instrument: str | None = None
    sequencing_method: str | None = None
    sequenced_at: datetime | None = None


class SampleBase(RWModel):
    """Base datamodel for sample data structure"""

    sample_id: str = Field(..., min_length=3, max_length=100)
    sample_name: str
    lims_id: str

    # metadata
    sequencing: SequencingInfo
    pipeline: PipelineRun

    # quality
    qc: list[Any] = Field(..., default_factory=list)

    # species identification
    species_prediction: list[Any] = Field(..., default_factory=list)



class AnalysisResult(BaseModel):
    """Container for analysis results."""
    software: str
    software_version: str

    # how it was parsed
    parser_name: str
    parser_version: int
    parser_status: str
    reason: str | None = None

    # results
    analysis_type: str
    results: Any


class GenericMetadataRecord(BaseModel):
    """Canonical internal representation of simple metadata values."""

    fieldname: str
    value: str | int | float | datetime
    category: str | None = None
    data_type: Literal["string", "integer", "float", "datetime"]


class TabularMetadataRecord(BaseModel):
    fieldname: str
    columns: list[str]
    rows: list[str] | None = None
    cells: list[list[Any]]
    category: str | None = None
    data_type: Literal["table"] = "table"


InternalMetadataRecord: TypeAdapter = GenericMetadataRecord | TabularMetadataRecord

class ParsedSampleResults(BaseModel):
    """Internal representation of a parsed sample."""
    schema_version: Literal[3] = 3

    sample_id: str
    sample_name: str
    lims_id: str

    # metadata
    metadata: list[InternalMetadataRecord] = Field(..., default_factory=list)
    sequencing: SequencingInfo
    pipeline: PipelineRun

    # analysis results
    analysis_results: list[AnalysisResult] = Field(..., default_factory=list)


class PipelineResult(SampleBase):
    """Input format of sample object from pipeline."""
    schema_version: Literal[2] = SCHEMA_VERSION

    # optional typing
    typing_result: list[MethodIndex] = Field(..., default_factory=list)
    # optional phenotype prediction
    element_type_result: list[MethodIndex] = Field(..., default_factory=list)
    # optional variant info
    snv_variants: list[VariantBase] | None = None
    sv_variants: list[VariantBase] | None = None
    indel_variants: list[VariantBase] | None = None
    # optional alignment info
    reference_genome: ReferenceGenome | None = None
    read_mapping: str | None = None
    genome_annotation: list[IgvAnnotationTrack] | None = None
